wicket-extensions/src/main/java/org/apache/wicket/extensions/markup/html/repeater/data/table/NavigationToolbar.java

```diff
@@ -44,17 +46,18 @@ public NavigationToolbar(final DataTable<?, ?> table)
 
 		WebMarkupContainer span = new WebMarkupContainer("span");
 		add(span);
-		span.add(AttributeModifier.replace("colspan", new IModel<String>()
-		{
-			@Override
-			public String getObject()
-			{
-				return String.valueOf(table.getColumns().size()).intern();
-			}
-		}));
+		span.add(AttributeModifier.replace("colspan", (IModel<String>) () -> String.valueOf(table.getColumns().size()).intern()));
 
 		span.add(newPagingNavigator("navigator", table));
-		span.add(newNavigatorLabel("navigatorLabel", table));
+		Component complexLabel = newComplexNavigatorLabel("navigatorLabel", table);
+		if (complexLabel != null)
+		{
+			span.add(complexLabel);
+		}
+		else
+		{
+			span.add(newNavigatorLabel("navigatorLabel", table));
```

Comment:
the `null` will be added here in case `newNavigatorLabel` is not overwritten and `complexLabel == null`

Reply:
Bot following you. We don't want to break API. New method will allows you to replace simple label with a come complex thing: this is not possible now because newNavigatorLabel  factory method returns a WebComponent and not a component. By default new method returns null. Thus existing applications will not be broken.

Reply:
I think it is backward compatible to change the return type from `X` to `AParentOfX`.
Try it! The `japicmp` plugin will tell you!

Reply:
@martin-g I did. It was the first thing I tried and the plugin complained. Thus I tried this.

Reply:
@martin-g for wicket 10.x maybe we can change return type for 9.x plugin will complain and project does not build.

Reply:
That's unfortunate! It is source compatible but not binary compatible.
Let's see what others think about it. 
I don't like the change but I see no other way to do it.
And we have the same problem for org.apache.wicket.extensions.ajax.markup.html.AjaxEditableChoiceLabel#newLabel()

Reply:
> That's unfortunate! It is source compatible but not binary compatible. Let's see what others think about it. I don't like the change but I see no other way to do it. And we have the same problem for org.apache.wicket.extensions.ajax.markup.html.AjaxEditableChoiceLabel#newLabel()

Ma neither. I first try to do as you suggested + mvn clean build and I build failed. Thus I had too opt for this HACKY ugly thing. :-( In wicket 10.x we can drop the new label. The use case here is I want to plug there a panel with pagination label + some text filter. TO do this I had to roll my own NavigationToolbar... not nice at all

Reply:
I understand.
But IMO we should not promote the new method in 9.x. I.e. we should not deprecate the old method.
Here is what I think would be the best:
1) Wicket 9.x - add the new method and use it. Just as in the PR but remove the deprecations. Also make it clear in the javadoc of the new method that it is there temporarily for 9.x
2) Wicket 10.x - just change the return type to `Component`

Reply:
I have done this for this branch

Reply:
I will create PR for wicket 10.x